controller/konnect/watch_kongsni.go

```diff
@@ -0,0 +1,92 @@
+package konnect
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+// KongSNIReconciliationWatchOptions returns the watch options for
+// the KongSNI.
+func KongSNIReconciliationWatchOptions(cl client.Client,
+) []func(*ctrl.Builder) *ctrl.Builder {
+	return []func(*ctrl.Builder) *ctrl.Builder{
+		func(b *ctrl.Builder) *ctrl.Builder {
+			return b.For(
+				&configurationv1alpha1.KongSNI{},
+				builder.WithPredicates(
+					predicate.NewPredicateFuncs(kongSNIRefersToKonnectGatewayControlPlane(cl)),
+				),
+			)
+		},
+
+		func(b *ctrl.Builder) *ctrl.Builder {
+			return b.Watches(
+				&configurationv1alpha1.KongCertificate{},
+				handler.EnqueueRequestsFromMapFunc(enqueueKongSNIForKongCertificate(cl)),
+			)
+		},
+	}
+}
+
+func kongSNIRefersToKonnectGatewayControlPlane(
+	cl client.Client) func(client.Object) bool {
+	return func(obj client.Object) bool {
+		sni, ok := obj.(*configurationv1alpha1.KongSNI)
+		if !ok {
+			return false
+		}
+
+		certNN := types.NamespacedName{
+			Namespace: sni.Namespace,
+			Name:      sni.Spec.CertificateRef.Name,
+		}
+		cert := configurationv1alpha1.KongCertificate{}
+		if err := cl.Get(context.Background(), certNN, &cert); client.IgnoreNotFound(err) != nil {
+			return true
+		}
+
+		return cert.Spec.ControlPlaneRef != nil && cert.Spec.ControlPlaneRef.Type == configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef
+	}
+}
+
+func enqueueKongSNIForKongCertificate(
+	cl client.Client) func(context.Context, client.Object) []reconcile.Request {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		cert, ok := obj.(*configurationv1alpha1.KongCertificate)
+		if !ok {
+			return nil
+		}
+
+		cpRef := cert.Spec.ControlPlaneRef
+		if cpRef == nil || cpRef.Type != configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef {
+			return nil
+		}
+
+		sniList := configurationv1alpha1.KongSNIList{}
+		if err := cl.List(ctx, &sniList, client.InNamespace(cert.Namespace)); err != nil {
```

Comment:
Can we add an index field in here so that we don't need to filter below?

Similarly as it's done for e.g. `KongCredentialACL` https://github.com/Kong/gateway-operator/blob/3486d362a663afbe47f5f1c1703334902093b59a/controller/konnect/watch_credentialacl.go#L194-L212

Reply:
Done in c649560c523cf06327609614cfaac88b1eb1946b